gpt_index/indices/query/query_runner.py

```diff
@@ -137,63 +153,130 @@ def _get_query_obj(
+            query_str_or_bundle, extra_info=transform_extra_info
+        )
+
+        # query
+        query_obj = self._get_query_obj(index_struct)
+        if self._recursive:
+            # call recursively
+            nodes = query_obj.retrieve(query_bundle)
+
+            # do recursion here
+            nodes_for_synthesis = []
+            additional_source_nodes = []
+            for node_with_score in nodes:
+                if isinstance(node_with_score.node, IndexNode):
+                    index_node = node_with_score.node
+                    # recursive call
+                    response = self.query(query_bundle, index_node.index_id)
+
+                    new_node = Node(
+                        text=str(response)
+                    )
+                    nodes_for_synthesis.append(new_node)
+                    additional_source_nodes.extend(response.source_nodes)
+            
+            return query_obj.synthesize(query_bundle, nodes_for_synthesis, additional_source_nodes)
+        else:
+            return query_obj.query(query_bundle)
 
-        return query_combiner.run(query_obj, query_bundle)
-        # return query_obj.query(query_bundle)
 
     async def aquery(
         self,
         query_str_or_bundle: Union[str, QueryBundle],
-        index_struct: IndexStruct,
+        index_id: Optional[str] = None,
     ) -> Response:
         """Run query."""
+        if isinstance(self._index_struct, CompositeIndex):
+            if index_id is None:
+                index_id = self._index_struct.root_id
+            index_struct = self._index_struct.all_index_structs[index_id]
+        else:
+            if index_id is not None:
+                raise ValueError('index_id should be used with composite graph')
+            index_struct = self._index_struct
+
         # NOTE: Currently, query transform is only run once
         # TODO: Consider refactor to support index-specific query transform
+        # TODO: support query combiner
+        # query_combiner = self._get_query_combiner(index_struct, query_transform)
+
+        # query transform
         query_transform = self._get_query_transform(index_struct)
         transform_extra_info = {"index_struct": index_struct}
         query_bundle = query_transform(
             query_str_or_bundle, extra_info=transform_extra_info
         )
         query_obj = self._get_query_obj(index_struct)
 
-        return await query_obj.aquery(query_bundle)
+        if self._recursive:
+            # call recursively
+            nodes = query_obj.retrieve(query_bundle)
+
+            # do recursion here
+            nodes_for_synthesis = []
+            additional_source_nodes = []
+            for node_with_score in nodes:
+                if isinstance(node_with_score.node, IndexNode):
+                    index_node = node_with_score.node
+                    # recursive call
+                    response = await self.aquery(query_bundle, index_node.index_id)
```

Comment:
I recently landed an async change that allows parallel calls to be made down nodes (e.g. if you compose a list index over 5 vector indexes, you can have parallel calls for each vector index). 

there will be inevitably merge conflicts with that change, but i'm fine using your version.

that said though could we run asyncio.gather on these nodes to replicate that behavior?

Reply:
Agreed on `asyncio.gather` to replicate behavior. Will take a look at the PR later and revise to match.